### ISSUE TITLE
openstack_networking_router: enable/disable snat

### ISSUE THREAD

**Issue by Crapworks** [terraform-providers/terraform-provider-openstack#134]
Hi there,

I learned today that some OpenStack providers have SNAT disabled by default. Of course, I'd like to enable it, so I can actually access the internet from instances without attached floating IPs...

The router resource doesn't support it, so I tried to add the argument, just to find out that this is not supported by Gophercloud. Do you think it is worth implementing? In this case, I could propose a PR for Gophercloud as well.

Cheers,
Christian

**Comment by jtopjian** [comment]
@Crapworks What's the missing parameter?

The `openstack_networking_router_v2` resource supports something we call `value_specs` which is a funny way of saying "arbitrary key/values". `value_specs` are used for vendor-specific JSON parameters, but it's also a workaround for injecting missing parameters from Gophercloud.

**Comment by Crapworks** [comment]
@jtopjian Yes, I know about that and it is very useful, but in this case, the parameter I need (`enable_snat` ) is inside the `external_gateway_info` structure. From the OpenStack API docs:

```json
{
    "router": {
        "name": "router1",
        "external_gateway_info": {
            "network_id": "ae34051f-aa6c-4c75-abf5-50dc9ac99ef3",
            "enable_snat": true,
            "external_fixed_ips": [
                {
                    "ip_address": "172.24.4.6",
                    "subnet_id": "b930d7f6-ceb7-40a0-8b81-a425dd994ccf"
                }
            ]
        },
        "admin_state_up": true
    }
}
```

And as far as I know I can't inject a parameter in a nested dict?

**Comment by jtopjian** [comment]
Ah, I see.

You're more than welcome to open a Gophercloud PR. Looks like a simple patch. Just remember to find where `enable_snat` is defined in the actual Neutron code. :)

**Comment by Crapworks** [comment]
Using my quite limited GO skills, I guess this is all that is needed to enable this? Looks a bit to easy for me...

https://github.com/gophercloud/gophercloud/pull/574

**Comment by jtopjian** [comment]
@Crapworks You're good to go if you want to add this to Terraform.

You'll have to do a vendor update. To do this, you'll need to install `govendor` and then:

```shell
$ cd ~/go/src/github.com/terraform-providers/terraform-provider-openstack
$ govendor fetch github.com/gophercloud/gophercloud/openstack/networking/v2/extensions/layer3/routers/...
$ git status # should only show the `routers` files being updated
$ git add .
$ git commit -m "vendor: Updating gophercloud
```

The key pieces being:

1. Make sure only the Gophercloud router files are updated
2. Have a dedicated commit

**Comment by jtopjian** [comment]
Implemented in #140
